lib/package.gi

```diff
@@ -159,6 +159,159 @@ InstallGlobalFunction( SetPackageInfo, function( record )
+        metapackagerecord.Dependencies.NeededOtherPackages := [];
+    fi;
+
+    name := subpackagerecord.PackageName;
+    version := subpackagerecord.Version;
+    version := Concatenation( "= ", version );
+
+    Add( metapackagerecord.Dependencies.NeededOtherPackages, [ name, version ] );
+
+end );
+
+#############################################################################
+##
+#F  EnrichSubpackageRecord( subpackagerecord, metapackagerecord )
+##
+
+BindGlobal( "EnrichSubpackageRecord", function( subpackagerecord, metapackagerecord )
+    if not IsBound( subpackagerecord.SubpackageOf ) then
+        subpackagerecord.SubpackageOf := metapackagerecord.PackageName;
+    fi;
+end );
+
+#############################################################################
+##
+#F  AddPackageInfo( files )
+##
+InstallGlobalFunction( AddPackageInfos, function( files, pkgdir, ignore, metapackagerecord... )
+    local file, record, pkgname, version, is_metapackage, name;
+    if Length( metapackagerecord ) = 1 then
+      is_metapackage := true;
+      metapackagerecord := metapackagerecord[ 1 ];
+    else
+      is_metapackage := false;
+    fi;
+
+    for file in files do
+      # Read the `PackageInfo.g' file.
+      Unbind( GAPInfo.PackageInfoCurrent );
+      Read( file[1] );
+      if IsBound( GAPInfo.PackageInfoCurrent ) then
+        record:= GAPInfo.PackageInfoCurrent;
+        Unbind( GAPInfo.PackageInfoCurrent );
+        pkgname:= LowercaseString( record.PackageName );
+        NormalizeWhitespace( pkgname );
+        version:= record.Version;
+
+        # If we have this version already then leave it out.
+        if ForAll( GAPInfo.PackagesInfo,
+                    r ->    r.PackageName <> record.PackageName
+                        or r.Version <> version ) then
+
+          # Check whether GAP wants to reset loadability.
+          if     IsBound( GAPInfo.PackagesRestrictions.( pkgname ) )
+              and GAPInfo.PackagesRestrictions.( pkgname ).OnInitialization(
+                      record ) = false then
+            Add( GAPInfo.PackagesInfoRefuseLoad, record );
+          elif pkgname in ignore then
+            LogPackageLoadingMessage( PACKAGE_DEBUG,
+                Concatenation( "ignore package ", record.PackageName,
+                " (user preference PackagesToIgnore)" ), "GAP" );
+          else
+            record.InstallationPath:= Filename( [ pkgdir ], file[2] );
+            if not IsBound( record.PackageDoc ) then
+              record.PackageDoc:= [];
+            elif IsRecord( record.PackageDoc ) then
+              record.PackageDoc:= [ record.PackageDoc ];
+            fi;
+
+            if IsBound( record.IsMetaPackage ) and record.IsMetaPackage = true then
+              for name in Set( DirectoryContents( record.InstallationPath ) ) do
+                files := FindPackageInfosInSubdirectories( Directory( record.InstallationPath ), name );
```

Comment:
Optional: extract `dir := Directory( record.InstallationPath );` here and in the next line?